proposals/NNNN-linalg-matrix.md

```diff
@@ -0,0 +1,110 @@
+<!-- {% raw %} -->
+
+# Linear Algebra Matrix
+
+* Proposal: [NNNN](NNNN-linalg-matrix.md)
+* Author(s): [Chris Bieneman](https://github.com/llvm-beanz)
+* Sponsor: TBD
+* Status: **Under Consideration**
+
+## Introduction
+
+GPUs are exceptional parallel data processors, but increasingly it is becoming
+important to model operations with cross-thread data dependencies. In HLSL and
+Direct3D these operations have been called Wave operations, or in Vulkan
+Subgroup operations. Related terms like Quad or derivatives have similar
+meaning in different scoping contexts. Vulkan has also recently introduced the
+term "cooperative" when talking about operations that require participation from
+multiple threads, these can be viewed much like derivative operations but across
+the full SIMD unit instead of a subset of threads.
+
+All of these terms refer to the way the underlying instructions execute, not
+necessarily what they do. One big part of this proposal is to take 5 steps back
+and talk about what they do: linear algebra.
+
+## Motivation
+
+HLSL has a Vulkan extension for SIMD matrix types [0021 - vk::Cooperative
+Matrix](0021-vk-coop-matrix.md), and DirectX had previewed a similar feature in
+SM 6.8 called [Wave Matrix](https://github.com/microsoft/hlsl-specs/pull/61).
+This proposal is aimed at merging the two into a unified language feature that
+can be supported on all platforms (with some platform-specific limitations).
+
+## Proposed solution
+
+Below is a proposed pseudo-HLSL API. The proposal uses C++20 concepts to
+represent template type constraints so as to avoid needing SFINAE complications.
+
+```c++
+namespace hlsl {
+
+template <class T>
+concept ArithmeticScalar = std::is_arithmetic<T>::value;
+
+namespace linalg {
+
+template <typename ComponentTy, uint M, uint N>
+  requires ArithmeticScalar<ComponentTy>
+class Matrix {
+  template <typename NewCompTy> Matrix<NewCompTy, M, N> cast();
+
+  Matrix operator+(Matrix);
+  Matrix operator-(Matrix);
+  Matrix operator*(Matrix);
+  Matrix operator/(Matrix);
+
+  template <typename T>
+    requires ArithmeticScalar<T>
+  Matrix operator+(T);
+  template <typename T>
+    requires ArithmeticScalar<T>
+  Matrix operator-(T);
+  template <typename T>
+    requires ArithmeticScalar<T>
+  Matrix operator*(T);
+  template <typename T>
+    requires ArithmeticScalar<T>
+  Matrix operator/(T);
+
+  static Matrix Splat(ElTy Val);
+  static Matrix Load(ByteAddressBuffer Res, uint StartOffset, uint Stride,
+                     bool ColMajor, uint Align = sizeof(ComponentTy));
+  static Matrix Load(RWByteAddressBuffer Res, uint StartOffset, uint Stride,
+                     bool ColMajor, uint Align = sizeof(ComponentTy));
+
+  static Matrix Load(groupshared ElTy Arr[], uint StartIdx, uint Stride,
+                     bool ColMajor);
+
+  void Store(RWByteAddressBuffer Res, uint StartOffset, uint Stride,

```

Comment:
Assuming all `Store`s are preceded by a Load, can we simplify the API a little bit by defaulting to whatever was provided in `Load` for everything but the storage location/offset? 

I suppose this doesn't work for a `Matrix` created by a matrix binop but it seems like things like Stride and ColMajor would be retained?

Reply:
Not necessarily. You may not be storing back to the same place you loaded from, or in the same orientation.